### ISSUE TITLE
Trilogic Expert Cartridge

### ISSUE THREAD

**Issue by billybicep** [GideonZ/1541ultimate#126]
Currently it isn't possible to use a .crt file of the trilogic expert cartridge. I have also tried a physical cartridge and this doesn't work either. Is this planned to be fixed?

**Comment by GideonZ** [comment]
Please attach the CRT file of the trilogic cartridge that you have tested, and describe the expected result.

**Comment by markusC64** [comment]
@GideonZ : Here are Manuals, Binary files for the ROM and so on: https://rr.pokefinder.org/wiki/Expert_Cartridge
That might help to understand the expected behavior.

**Comment by markusC64** [comment]
And this does look like the missing crt: http://www.planetemu.net/rom/commodore-c64-applications-crt/expert-cartridge-1986-trilogic

**Comment by billybicep** [comment]
[expert_cartridge.zip](https://github.com/GideonZ/1541ultimate/files/4054638/expert_cartridge.zip)
[Expert Cartridge (1986)(Trilogic).zip](https://github.com/GideonZ/1541ultimate/files/4054642/Expert.Cartridge.1986.Trilogic.zip)

here's 2 .crt images.

**Comment by billybicep** [comment]
The Expert was a little different to other freezer carts as you had to program it each time when powering on the c64. It had a 3 way switch, PRG - OFF - ON. And a reset button.
You'd switch the cart to PRG then load up the software, the software would then say 'Switch to ON and press Reset' this would then transfer the loaded up software into the cart and it would stay active until you turned off the C64.

to freeze you then just pressed the Restore key, this would then go straight into a monitor/assemble with lots of commands available for saving, freezing, coding etc.

The 2 images work in VICE (not sure if this is relevant?)

**Comment by billybicep** [comment]
If i try to load the .crt on the U64 it just says 'Expert cartridge not implemented' . If i try a physical expert cart it just hangs when you move it to ON.

**Comment by billybicep** [comment]
I know Expert wasnt as popular as Action Replay etc but it was very popular in the UK for hacking and coding, it had an excellent assembler written by J. Twiddy

**Comment by GideonZ** [comment]
Ahhhh... Okay, now I understand why you said the CRT file wouldn't work. I already found that very peculiar. It is simply not implemented. :-D  (Which is not the same as: "It doesn't work..")

The external cart is supposed to work, however. I can see in the schematics that the cart is somewhat 'dirty' in trying to push the NMI line high in some situations, overruling the CIA and restore key. I think that's not very nice and desirable. I'd suggest that an implementation of the Trilogic cart in the FPGA is a way better solution, albeit that it that the external cart should at least work for loading and showing up the menu. Apparently it doesn't and this indeed exposes a bug or timing problem in the U64 bus logic or PLA.

As for the FPGA implementation of the TriLogic cart, I suggest that the CRT file is simply loaded into RAM, and after loading, the "switch" would be in the ON mode. There is no need for the PGM mode. In addition, the freeze function of the U64 can be simply routed to this bit of logic as well. May need some refactoring, as the current freeze function always forces both NMI and IRQ low, like all other freezers do. So the freeze debouncing and flipflop need to be separated from its output. Maybe that is already so. Not being able to push the NMI line high, might actually limit the usability of the cart, as NMI is often used for timers and such.

**Comment by billybicep** [comment]
I agree that it would be much better in the FPGA, i only tried the physical cart as i was keen to use expert again!

**Comment by GideonZ** [comment]
I'll see if I can obtain one of those carts, so I can do some tests with it.

**Comment by billybicep** [comment]
I have a spare I can send to you if thats easier?

**Comment by GideonZ** [comment]
That'd be highly appreciated. One thing to make clear: I do not make
promises of WHEN I can take a look at it - the list of issues is
already quite long.

**Comment by billybicep** [comment]
Thats fine, i understand you must have a lot of requests/issues. I'll send it to the address on your website.

**Comment by billybicep** [comment]
Expert cartridge has been posted, hopefully should be with you soon.

**Comment by GideonZ** [comment]
Great!  Thank you very much for the update!
Seems like I am going to be really really busy!

**Comment by billybicep** [comment]
[Expert 210.zip](https://github.com/GideonZ/1541ultimate/files/4133867/Expert.210.zip)

Expert v2.10 software, this was probably the most stable version they did.
